File: src/main/java/de/aivot/GoverBackend/elements/models/elements/form/input/FunctionInputElement.java
package de.aivot.GoverBackend.elements.models.elements.form.input;

import de.aivot.GoverBackend.core.services.ObjectMapperFactory;
import de.aivot.GoverBackend.elements.models.elements.BaseInputElement;
import de.aivot.GoverBackend.enums.ElementType;
import de.aivot.GoverBackend.exceptions.RequiredValidationException;
import de.aivot.GoverBackend.exceptions.ValidationException;
import jakarta.annotation.Nonnull;
import jakarta.annotation.Nullable;

import java.time.format.DateTimeFormatter;
import java.util.Objects;

public class FunctionInputElement extends BaseInputElement<FunctionInputElementItem> {
    @Nonnull
    private FunctionReturnType returnType = FunctionReturnType.BOOLEAN;

    public FunctionInputElement() {
        super(ElementType.FunctionInput);
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        FunctionInputElement that = (FunctionInputElement) o;
        return returnType == that.returnType;
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), returnType);
    }

    @Override
    public FunctionInputElementItem formatValue(Object value) {
        if (value == null) {
            return null;
        }

        return ObjectMapperFactory
                .getInstance()
                .convertValue(value, FunctionInputElementItem.class);
    }

    @Override
    public void performValidation(@Nullable FunctionInputElementItem value) throws ValidationException {
        if (value == null) {
            if (Boolean.TRUE.equals(getRequired())) {
                throw new RequiredValidationException(this);
            }
        }
    }

    @Nonnull
    public FunctionReturnType getReturnType() {
        return returnType;
    }

    public FunctionInputElement setReturnType(FunctionReturnType returnType) {
        this.returnType = Objects.requireNonNullElse(returnType, FunctionReturnType.BOOLEAN);
        return this;
    }

    public enum FunctionReturnType {
        STRING,
        NUMBER,
        BOOLEAN,
        DATE,
        DATETIME
    }
}

File: app/src/models/elements/form/input/function-input-element.ts
import {BaseInputElement} from '../base-input-element';
import {ElementType} from '../../../../data/element-type/element-type';
import {JavascriptCode} from "../../../functions/javascript-code";
import {ConditionSet} from "../../../functions/conditions/condition-set";
import {NoCodeOperand} from "../../../functions/no-code-expression";

export interface FunctionInputElement extends BaseInputElement<ElementType.FunctionInput> {
    returnType: FunctionInputElementReturnType | null;
}

export enum FunctionInputElementReturnType {
    STRING = 'STRING',
    NUMBER = 'NUMBER',
    BOOLEAN = 'BOOLEAN',
    DATE = 'DATE',
    DATETIME = 'DATETIME',
}

export interface FunctionInputElementItem {
    type: FunctionInputElementItemType | null;
    code: JavascriptCode | null;
    conditionSet: ConditionSet | null;
    noCode: NoCodeOperand | null;
}

export enum FunctionInputElementItemType {
    NoCode = 'NoCode',
    ConditionSet = 'ConditionSet',
    Javascript = 'Javascript',
}